### PR TITLE
feat(auth): enables OIDC auth code flow

### PR DESCRIPTION
**Pull request by ScruffyProdigy**
Provides an option for developers to specify the OAuth response type for their OIDC provider (either one of these can be set:):

- id_token
- code (if set, must also set the client secret)

RELEASE NOTES: Added support for configuring the authorization code flow for OIDC providers.

### CODE REVIEW

**Comment by ScruffyProdigy** [comment]
@lsirac

File: tests/test_auth_providers.py [hunk 1]
```diff
@@ -112,6 +120,9 @@ def test_get(self, user_mgt_app):
         {'issuer': None}, {'issuer': ''}, {'issuer': 'not a url'},
         {'display_name': True},
         {'enabled': 'true'},
+        {'id_token_response_type': 'true'}, {'code_response_type': 'true'},
+        {'code_response_type': True, 'client_secret': ''}, {'code_response_type': True, 'client_secret': True},
+        {'code_response_type': True, 'client_secret': None},
```
Comment (hiranya911) [inline tests/test_auth_providers.py:125]: Also prevent setting both to `False`?

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Looks mostly good. Needs a few tweaks to the public API, and a couple of readability nits.

File: firebase_admin/_auth_providers.py [hunk 2]
```diff
@@ -59,6 +59,26 @@ def issuer(self):
     def client_id(self):
         return self._data['clientId']
 
+    @property
+    def client_secret(self):
+        return self._data.get('clientSecret')
+
+    @property
+    def response_type(self):
```
[the remark this answers is not in the document]
Comment (ScruffyProdigy) [reply firebase_admin/_auth_providers.py:67]: Done

File: firebase_admin/_auth_providers.py [hunk 3]
```diff
@@ -59,6 +59,26 @@ def issuer(self):
...
+        return ResponseTypeProviderConfig(self._data['responseType'])
+
+
+class ResponseTypeProviderConfig(ProviderConfig):
```
[the remark this answers is not in the document]
Comment (ScruffyProdigy) [reply firebase_admin/_auth_providers.py:71]: Done

File: firebase_admin/_auth_providers.py [hunk 4]
```diff
@@ -190,13 +211,23 @@ def create_oidc_provider_config(
             req['displayName'] = _auth_utils.validate_string(display_name, 'display_name')
         if enabled is not None:
             req['enabled'] = _auth_utils.validate_boolean(enabled, 'enabled')
+        if id_token_response_type is not None or code_response_type is not None:
```
[the remark this answers is not in the document]
Comment (ScruffyProdigy) [reply firebase_admin/_auth_providers.py:214]: Done

File: firebase_admin/_auth_providers.py [hunk 5]
```diff
@@ -211,6 +242,14 @@ def update_oidc_provider_config(
             req['clientId'] = _validate_non_empty_string(client_id, 'client_id')
         if issuer:
             req['issuer'] = _validate_url(issuer, 'issuer')
+        if id_token_response_type is not None or code_response_type is not None:
```
[the remark this answers is not in the document]
Comment (ScruffyProdigy) [reply firebase_admin/_auth_providers.py:245]: Done

File: integration/test_auth.py [hunk 6]
```diff
@@ -692,6 +692,9 @@ def test_create_oidc_provider_config(oidc_provider):
     assert oidc_provider.issuer == 'https://oidc.com/issuer'
     assert oidc_provider.display_name == 'OIDC_DISPLAY_NAME'
     assert oidc_provider.enabled is True
+    assert oidc_provider.response_type.id_token == True
```
[the remark this answers is not in the document]
Comment (ScruffyProdigy) [reply integration/test_auth.py:695]: Done

File: integration/test_auth.py [hunk 7]
```diff
@@ -692,6 +692,9 @@ def test_create_oidc_provider_config(oidc_provider):
...
+    assert oidc_provider.response_type.code == False
+    assert oidc_provider.client_secret == None
```
[the remark this answers is not in the document]
Comment (ScruffyProdigy) [reply integration/test_auth.py:697]: Done

File: tests/test_auth_providers.py (hunk 1 above)
Comment (ScruffyProdigy) [reply tests/test_auth_providers.py:125]: Done

File: public/index.html [hunk 8]
```diff
@@ -0,0 +1,89 @@
+<!DOCTYPE html>
```
Comment (hiranya911) [inline public/index.html:1]: Please remove :)

File: firebase_admin/_auth_providers.py [hunk 9]
```diff
@@ -220,6 +212,18 @@ def create_oidc_provider_config(
                 if code_response_type:
                     req['clientSecret'] = _validate_non_empty_string(client_secret, 'client_secret')
 
+        response_type = {}
+        if id_token_response_type is False and code_response_type is False:
+            raise ValueError("At least one response type must be returned.")
```
Comment (hiranya911) [inline firebase_admin/_auth_providers.py:217]: Please use single quotes for string literals.

File: firebase_admin/_auth_providers.py [hunk 10]
```diff
@@ -251,6 +255,18 @@ def update_oidc_provider_config(
                     req['clientSecret'] = _validate_non_empty_string(client_secret, 'client_secret')
                 req['responseType']['code'] = _auth_utils.validate_boolean(code_response_type, 'code_response_type')
 
+        response_type = {}
+        if id_token_response_type is False and code_response_type is False:
+            raise ValueError("At least one response type must be returned.")
```
Comment (hiranya911) [inline firebase_admin/_auth_providers.py:260]: Single quotes here too.

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Looks mostly good. Please fix the errors reported by the linter. Plus couple of minor nits on style.

File: tests/test_auth_providers.py [hunk 11]
```diff
@@ -198,7 +225,7 @@ def test_update(self, user_mgt_app):
         assert len(recorder) == 1
         req = recorder[0]
         assert req.method == 'PATCH'
-        mask = ['clientId', 'displayName', 'enabled', 'issuer']
+        mask = ['clientId', 'clientSecret', 'displayName', 'enabled', 'issuer', 'responseType.code', 'responseType.idToken']
```
Comment (lahirumaramba) [inline tests/test_auth_providers.py:228]: The following lines also seem to be triggering more lint errors.
```
Running linter on module tests
************* Module tests.test_auth_providers
tests/test_auth_providers.py:124:0: C0301: Line too long (111/100) (line-too-long)
tests/test_auth_providers.py:208:0: C0301: Line too long (111/100) (line-too-long)
tests/test_auth_providers.py:228:0: C0301: Line too long (124/100) (line-too-long)
tests/test_auth_providers.py:253:0: C0301: Line too long (127/100) (line-too-long)
```

**Comment by lahirumaramba** [comment]
Thanks @ScruffyProdigy !
Hi @egilmorez please take a look at the reference docs when you get a chance. Thank you!

**Review by hiranya911 — APPROVED** [review]
Thanks. LGTM 👍

**Review by lahirumaramba — APPROVED** [review]

File: firebase_admin/_auth_client.py [hunk 12]
```diff
@@ -528,6 +529,15 @@ def create_oidc_provider_config(
                 This name is also used as the provider label in the Cloud Console.
             enabled: A boolean indicating whether the provider configuration is enabled or disabled
                 (optional). A user cannot sign in using a disabled provider.
+            client_secret: A string which sets the client secret for the new provider.
+                This is required for the code flow.
+            code_response_type: Sets whether to enable the code response flow for the new provider.
```
Comment (kevinthecheung) [inline firebase_admin/_auth_client.py:534]: Is this a boolean?

File: firebase_admin/_auth_client.py [hunk 13]
```diff
@@ -528,6 +529,15 @@ def create_oidc_provider_config(
...
+                By default, this is not enabled if no response type is specified.
+                A client secret must be set for this response type.
+                Having both the code and ID token response flows is currently not supported.
+            id_token_response_type: Sets whether to enable the ID token response flow for the new
```
Comment (kevinthecheung) [inline firebase_admin/_auth_client.py:538]: Is this a boolean?

**Review by kevinthecheung — CHANGES_REQUESTED** [review]

**Review by egilmorez — APPROVED** [review]
LGTM!

File: firebase_admin/_auth_client.py (hunk 12 above)
Comment (ScruffyProdigy) [reply firebase_admin/_auth_client.py:534]: Yes. Fixed

File: firebase_admin/_auth_client.py (hunk 13 above)
Comment (ScruffyProdigy) [reply firebase_admin/_auth_client.py:538]: Yes. Fixed

**Review by kevinthecheung — APPROVED** [review]

**Pull request merged**